File: common/src/main/java/dev/obscuria/tooltips/client/label/RarityLabelProvider.java
package dev.obscuria.tooltips.client.label;

import com.mojang.serialization.Codec;
import net.minecraft.ChatFormatting;
import net.minecraft.client.gui.screens.inventory.tooltip.ClientTooltipComponent;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Rarity;

public record RarityLabelProvider() implements LabelProvider {

    public static final RarityLabelProvider INSTANCE = new RarityLabelProvider();
    public static final Codec<RarityLabelProvider> CODEC = Codec.unit(INSTANCE);

    @Override
    public Codec<RarityLabelProvider> codec() {
        return CODEC;
    }

    @Override
    public ClientTooltipComponent create(ItemStack stack) {
        final var key = makeRarityKey(stack.getRarity());
        final var component = Component.translatable(key).withStyle(ChatFormatting.GRAY);
        return ClientTooltipComponent.create(component.getVisualOrderText());
    }

    private String makeRarityKey(Rarity rarity) {
        final var name = rarity.name().toLowerCase().replaceAll(":", ".");
        return "rarity." + name + ".name";
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/label/RarityLabelProvider.kt
package dev.obscuria.tooltips.client.label

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.ChatFormatting
import net.minecraft.client.gui.screens.inventory.tooltip.ClientTooltipComponent
import net.minecraft.network.chat.Component
import net.minecraft.world.item.ItemStack

class RarityLabelProvider(
    val drawBackground: Boolean
) : ILabelProvider {

    override fun codec(): Codec<RarityLabelProvider> = CODEC

    override fun bake(stack: ItemStack): ClientTooltipComponent {
        val component = Component
            .translatable("rarity.${stack.rarity.name.lowercase()}")
            .withStyle(ChatFormatting.GRAY)
        return ClientTooltipComponent.create(component.visualOrderText)
    }

    companion object {

        val CODEC: Codec<RarityLabelProvider> = RecordCodecBuilder.create { codec ->
            codec.group(
                Codec.BOOL
                    .optionalFieldOf("draw_background", false)
                    .forGetter(RarityLabelProvider::drawBackground)
            ).apply(codec, ::RarityLabelProvider)
        }
    }
}